### PR TITLE
feat: add blazor template

### PR DESCRIPTION
**Pull request by AerafalDev**
add blazor template for create tauri app with blazor

[Demo](https://github.com/tauri-apps/create-tauri-app/assets/126319968/6029ad0a-0acb-478a-befe-662e9296fa6b)

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -420,6 +425,59 @@ fn to_valid_pkg_name(project_name: &str) -> String {
     }
 }
 
+fn create_dotnet_webassembly_project(project_name: &str, target_dir: &std::path::PathBuf) -> anyhow::Result<()> {
```
Comment (amrbashir) [inline src/lib.rs]: The template shouldn't be bootstrapped using the dotnet CLI and instead should be added to `templates` directory in this repo same as other templates. The template should replicate this look and behavior 
image

**Review by amrbashir — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (AerafalDev) [reply src/lib.rs]: I'm going to take care of appearance and behavior. Indeed, at first I wanted to do as you just told me but I encountered a problem, in C# we have .csproj files that serve as a repository for our project, that said, I couldn't rename the .csproj with the project nameIndeed, at the beginning I wanted to do as you just told me but I encountered a problem, in C# we have .csproj files that serve as a repository for our project, that said, I couldn't rename the .csproj with the project name, y to Is there a way to change the name of a file with the name of the project like this ?
image

Comment (AerafalDev) [reply src/lib.rs]: I've just finished making the template, I'm still having the problem renaming my .csproj file

image

Comment (zenitogr) [reply src/lib.rs]: > I'm going to take care of appearance and behavior. Indeed, at first I wanted to do as you just told me but I encountered a problem, in C# we have .csproj files that serve as a repository for our project, that said, I couldn't rename the .csproj with the project nameIndeed, at the beginning I wanted to do as you just told me but I encountered a problem, in C# we have .csproj files that serve as a repository for our project, that said, I couldn't rename the .csproj with the project name, y to Is there a way to change the name of a file with the name of the project like this ? image

I think this logic needs to go in one of the .rs files in the src, I think template.rs?

Comment (amrbashir) [reply src/lib.rs]: @zenitogr after these lines https://github.com/tauri-apps/create-tauri-app/blob/96bb2a4477ef099a0c37bcd5b95b87045b3df16d/src/template.rs#L380-L387 you can add `let file_name = lte::render(file_name, manifest_template_data)?;`

but you need to name the file as `{% project_name %}.csproj` not `%{ project_name }%.csproj`, you have the syntax inverted

Comment (Aerafaal) [reply src/lib.rs]: thank you, I will do that today

Comment (AerafalDev) [reply src/lib.rs]: That's okay, I'm done everything is done according to your requirements @amrbashir

File: templates/template-blazor/src/wwwroot/index.html.lte [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>{% project_name %}</title>
```
Comment (amrbashir) [inline templates/template-blazor/src/wwwroot/index.html.lte]:
```suggestion
    <title>Tauri + Blazor</title>
```
and let's rename this file to just `index.html`

File: templates/template-blazor/src/wwwroot/css/app.css [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    transition: 0.75s;
+}
+
+.logo.tauri:hover {
+    filter: drop-shadow(0 0 2em #24c8db);
+}
+
+.logo.blazor:hover {
+    filter: drop-shadow(0 0 2em #5c2d91);
+}
```
Comment (amrbashir) [inline templates/template-blazor/src/wwwroot/css/app.css]: you only need to keep this bit here and remove everything else, the rest will be uplied from `_assets_/styles.css`

File: templates/template-blazor/.manifest [hunk 4]
```diff
@@ -0,0 +1,5 @@
+beforeDevCommand = dotnet watch run --project src/{% project_name %}.csproj
+beforeBuildCommand = dotnet publish -c release src/{% project_name %}.csproj -o dist
+devPath = http://localhost:1420
+distDir = ../dist/wwwroot
+withGlobalTauri = true
```
Comment (amrbashir) [inline templates/template-blazor/.manifest]:
```suggestion
withGlobalTauri = true

[files]
styles.css = src/wwwroot/css/app.css
tauri.svg = src/wwwroot/img/tauri.svg
```

File: templates/template-blazor/src/wwwroot/img/favicon.png
Comment (amrbashir) [inline templates/template-blazor/src/wwwroot/img/favicon.png:1]: favicons don't display on tauri apps, so let's remove this file

File: templates/template-blazor/src/wwwroot/index.html.lte [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    <base href="/" />
+    <link rel="stylesheet" href="css/app.css" />
+    <link rel="icon" type="image/png" href="img/favicon.png" />
```
Comment (amrbashir) [inline templates/template-blazor/src/wwwroot/index.html.lte]: and remove this favicon link as well

File: templates/template-blazor/src/wwwroot/img/tauri.svg
Comment (amrbashir) [inline templates/template-blazor/src/wwwroot/img/tauri.svg:1]: no need to re-add this file here, it will be copied over from `_assets_/tauri.svg`

**Review by amrbashir — CHANGES_REQUESTED** [review]
We also need to add a new category in `categories.rs` file

**Comment by AerafalDev** [comment]
> We also need to add a new category in `categories.rs` file

Does this imply creating a new package manager as well?
Because if I add a new category named CSharp, I end up with Cargo templates.
And since Tauri's CLI cannot be used in C#, I won't have a choice but to use Cargo.
So if I'm not mistaken, I'll create a new package manager named NuGet, but it will use Cargo?

Regarding the rest, everything is finished, only this point needs to be clarified.

**Comment by amrbashir** [comment]
> Does this imply creating a new package manager as well?
> Because if I add a new category named CSharp, I end up with Cargo templates.
> And since Tauri's CLI cannot be used in C#, I won't have a choice but to use Cargo.
> So if I'm not mistaken, I'll create a new package manager named NuGet, but it will use Cargo?

yeah it is a bit weird, let's go with `.NET` as the category name and `.NET CLI` as the package manager name.

Ideally we would publish the Tauri CLI to NuGet and users can call it using `dotnet tool run tauri`, is that even possible?

**Comment by AerafalDev** [comment]
> yeah it is a bit weird, let's go with `.NET` as the category name and `.NET CLI` as the package manager name.
> 
> Ideally we would publish the Tauri CLI to NuGet and users can call it using `dotnet tool run tauri`, is that even possible?

Yes this is possible, I'm going to do a poc, before that, what commands are required? Just the command to launch Tauri?

**Comment by amrbashir** [comment]
It should be able to run any tauri subcommand `tauri dev`, `tauri build`, `tauri info`...etc

**Comment by AerafalDev** [comment]
> It should be able to run any tauri subcommand `tauri dev`, `tauri build`, `tauri info`...etc

Alright, I just have to make a wrapper, which will implicitly call cargo in this case

**Comment by amrbashir** [comment]
> > It should be able to run any tauri subcommand `tauri dev`, `tauri build`, `tauri info`...etc
> 
> Alright, I just have to make a wrapper, which will implicitly call cargo in this case

No need for a wrapper, I was wondering if it is possible, we could publish tauri-cli to NuGet and then this template could include it as a dependency

**Comment by AerafalDev** [comment]
> No need for a wrapper, I was wondering if it is possible, we could publish tauri-cli to NuGet and then this template could include it as a dependency

publish tauri-cli on Nuget? But tauri-cli is made in rust, isn't it?

**Comment by amrbashir** [comment]
yeah, can't we publish a prebuilt binary there?

**Comment by amrbashir** [comment]
anyways, let's just add these categories for now and we can see about publishing tauri-cli to nuget later.

**Comment by amrbashir** [comment]
also make sure to update https://github.com/tauri-apps/create-tauri-app/blob/dev/.scripts/generate-templates-matrix.js and https://github.com/tauri-apps/create-tauri-app/blob/dev/.github/workflows/templates-test.yml so the new template can be tested in CI

**Comment by AerafalDev** [comment]
> yeah, can't we publish a prebuilt binary there?

No, we can only publish on Nuget the projects of the .NET suite (including C# and F#), binaries are not taken into account, that's why I thought I was more on the idea of making a wrapper in C#, to see later if you have a better idea than mine.

> also make sure to update https://github.com/tauri-apps/create-tauri-app/blob/dev/.scripts/generate-templates-matrix.js and https://github.com/tauri-apps/create-tauri-app/blob/dev/.github/workflows/templates-test.yml so the new template can be tested in CI

I'll take care of it right away.

**Comment by AerafalDev** [comment]
Everything is good for me @amrbashir

**Comment by AerafalDev** [comment]
https://github.com/jimmy-mll/create-tauri-app/actions/runs/8547370990/job/23419445341#step:17:14

I have a problem when building the job with the ci, the '-' character of 'tauri-app' is not supported in C# uses

**Comment by amrbashir** [comment]
> I have a problem when building the job with the ci, the '-' character of 'tauri-app' is not supported in C# uses

then we need to replace with a supported character it in these places

**Review by zenitogr — CHANGES_REQUESTED** [review]
Looks good, but, is there a bundler for tauri? I think using withGlobalTauri is not for release, excuse if this is incorrect i am a new dev myself

**Comment by AerafalDev** [comment]
https://github.com/jimmy-mll/create-tauri-app/actions/runs/8554756333

@amrbashir

**Review by Aerafaal — APPROVED** [review]

**Comment by amrbashir** [comment]
looks like you need to run `pnpm format` and `cargo fmt`

**Comment by AerafalDev** [comment]
> looks like you need to run `pnpm format` and `cargo fmt`

👍

**Comment by AerafalDev** [comment]
It's all good for me, I think we're not too bad here @amrbashir

File: .github/workflows/templates-test.yml [hunk 6]
```diff
@@ -121,18 +127,18 @@ jobs:
...
 
       - run: ${{ matrix.settings.install_cmd }}
         if: matrix.settings.install_cmd  != ''
-        working-directory: tauri-app
+        working-directory: ${{ (matrix.settings.manager == 'dotnet') && 'TauriApp' || 'tauri-app' }}
 
       - run: ${{ matrix.settings.run_cmd }} tauri build ${{ matrix.settings.no_bundle_flag }}
         if: matrix.settings.manager != 'npm'
-        working-directory: tauri-app
+        working-directory: ${{ (matrix.settings.manager == 'dotnet') && 'TauriApp' || 'tauri-app' }}
```
Comment (amrbashir) [inline .github/workflows/templates-test.yml]: when bootstrapping a dotnet project it should support `tauri-app` too, we shouldn't handle it here at all, we just need to convert `tauri-app` to `TauriApp` automatically in in the template in the files that require it to be without `-`, I guess for example that would be `Program.cs.lte`, and in that file we change `{% project_name %}` to `{% project_name_upper_case %}`. We will need to inject that variable into the template data though.

File: templates/template-blazor/src/Program.cs.lte [hunk 7]
```diff
@@ -0,0 +1,11 @@
+using Microsoft.AspNetCore.Components.Web;
+using Microsoft.AspNetCore.Components.WebAssembly.Hosting;
+using {% project_name %};
```
Comment (amrbashir) [inline templates/template-blazor/src/Program.cs.lte]: I think this one

```suggestion
using {% project_name_uppercase %};
```

File: templates/template-blazor/src/_Imports.razor.lte [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿@using System.Net.Http
+@using System.Net.Http.Json
+@using Microsoft.AspNetCore.Components.Forms
+@using Microsoft.AspNetCore.Components.Routing
+@using Microsoft.AspNetCore.Components.Web
+@using Microsoft.AspNetCore.Components.Web.Virtualization
+@using Microsoft.AspNetCore.Components.WebAssembly.Http
+@using Microsoft.JSInterop
+@using {% project_name %}
```
Comment (amrbashir) [inline templates/template-blazor/src/_Imports.razor.lte]: and this one
```suggestion
@using {% project_name_uppercase %}
```

File: templates/template-blazor/src/{% project_name %}.csproj
Comment (amrbashir) [inline templates/template-blazor/src/{% project_name %}.csproj:1]: and the name of this file

**Review by amrbashir — CHANGES_REQUESTED** [review]
Sorry for the delay, I left a few final comments and this should be ready to go.

**Comment by AerafalDev** [comment]
Normally, this time, it’s the right one @amrbashir
https://github.com/jimmy-mll/create-tauri-app/actions/runs/8620215358/job/23626733542

**Review by amrbashir — DISMISSED** [review]

**Review by amrbashir — APPROVED** [review]

**Pull request merged**